CC: CMake Tutorial and Reference Resources (#1259)

bartlettroscoe opened the pull request:
# Description

EB Member: @`<eb-member-id>`

Resolves: 
* #1259

This turned out to be a little longer than I had expected but I think it contains some important information that makes this CC article more useful just a set of links to CMake documentation resources.

NOTE: I included a image for the book "Professional CMake" but I am not sure it is needed and it lengthens the article.  But I guess we would keep it and move it to the bottom like for the CC article:

* https://bssw.io/items/digging-deeper-into-agile-software-development

## PR checklist for files displayed on bssw.io site

* [ ] ***[Author]*** `@mention` the BSSw.io editorial board member `@<eb-member-id>` in **Description** above assigned to shepherd your PR.
* [x] ***[Author]*** Add the `<issue-id>` in the **Description** above for the associated GitHub Issue.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py])~~.
* [x] ***[Author]*** Inspect the content in the `*.md` file(s) as rendered in GitHub for this PR.
* [ ] ***[EB Mem]*** Assign this PR to the EB member `<eb-member-id>`.
* [x] ***[EB Mem]*** Assign this PR to the author of the PR `<pr-author-id>`.
* [x] ***[EB Mem]*** Add one or more Reviewers.
* [x] ***[EB Mem]*** Add label `content: <content-type>` for the type of contribution.
* [x] ***[EB Mem]*** Add to Project `Content Development` (see [Content Development]).
* [x] ***[EB Mem]*** Add [meta-data] to the `*.md` file(s) (set `Publish: yes`).
* [x] ***[EB Mem]*** Add label `preview` (so PR branch will be merged to 'preview' branch and watch for possible merge failures).
* [x] ***[EB Mem]*** Rebuild [preview] site and confirm new content is there, renders correctly and is returned in searches.
* [x] ***[Author]*** Make any final changes to the PR based on feedback.
* ~~[ ] ***[Author]*** Ensure `wikize_refs.py -i <base>.md` is run and commit (if using [wikize_refs.py]).~~
* [x] ***[EB Mem]*** Rebuild [preview] site and re-confirm content looks correct.
* [x] ***[EB Mem]*** Ensure at least one reviewer signs off on the final changes.
* [x] ***[EB Mem]*** Move the PR to "Ready to Publish" in [Content Development].
* [x] ***[EB Mem]*** Leave comment and @mention `@betterscientificsoftware/bssw-maint` (**BSSw Maint**) asking to carry out final publication steps.
* [ ] ***[BSSw Maint]*** Verify that all needed files are present in the PR (article, images, updates to Site/Homepage.md carousel and/or Site/Announcements/Announcements.md as appropriate).
* [ ] ***[BSSw Maint]*** Merge PR. (Should automatically move to "Done" in [Content Development].)
* [ ] ***[BSSw Maint]*** Verify new contribution shows up on [bssw.io] as expected.

NOTE:
* Checklist items prefixed with ***[Author]*** are expected to be performed by the author of the PR or can be performed by the author.
* Checklist items prefixed with ***[EB Mem]*** must be performed by a [BSSw.io Editorial Board (EB) Member](https://betterscientificsoftware.github.io/bssw.io/bssw_members.html).
* Checklist items prefixed with ***[BSSw Maint]*** must be performed by someone in the `@betterscientificsoftware/bssw-maint` team (hint: type `@`, `b`, `s`, `s`, `w`, `-`, `m`  to auto-complete to `@betterscientificsoftware/bssw-maint`).


<!-- Standard links below, leave these this section! -->

[preview]: https://preview.bssw.io
[bssw.io]: https://bssw.io
[Content Development]: https://github.com/betterscientificsoftware/bssw.io/projects/3
[BSSw Internal]: https://github.com/betterscientificsoftware/bssw.io/projects/2
[meta-data]: https://betterscientificsoftware.github.io/bssw.io/bssw_styling_common.html#metadata-section
[wikize_refs.py]: https://github.com/betterscientificsoftware/bssw.io/blob/master/utils/README.md#wikize_refspy

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+
+<!--deck text start-->
+The CMake meta-build system is a large and powerful set of tools to create cross-platform builds from source for C, C++, Fortran and many other compiled programming languages.
+CMake continues to rapidly evolve and best practices can change from year to year (and even from release to release in some cases).
```
```suggestion
CMake continues to rapidly evolve and best practices in its use are changing almost as rapidly.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+<!--deck text start-->
+The CMake meta-build system is a large and powerful set of tools to create cross-platform builds from source for C, C++, Fortran and many other compiled programming languages.
+CMake continues to rapidly evolve and best practices can change from year to year (and even from release to release in some cases).
+Here we provide links and descriptions for the best tutorial and reference materials that exist for CMake to help stay on the forefront of effective CMake usage.
```
```suggestion
Here we provide links and descriptions for the best tutorial and reference materials that exist for CMake to help stay current in effective CMake usage.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+Publication | Year 2022, Self published, currently not versioned
+Web links | https://crascit.com/professional-cmake/
+
+The CMake meta-build tool and ecosystem (which includes the tools CTest and CPack with each CMake installation) continues to evolve and improve fairly rapidly from release to release.
```
Does this expository text apply to the one item *just above*, to all of the items above or to the item coming next? Greater verticle spacing or a horizontal linke (in GFM thats a sep. line with just `---`) might help to make this clear

bartlettroscoe replied:
All of the above.  So you suggest breaking out the intro with  `---`?  I will try that and see how it looks.

bartlettroscoe replied:
Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

bartlettroscoe replied:
Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+Web links | https://crascit.com/professional-cmake/
+
+The CMake meta-build tool and ecosystem (which includes the tools CTest and CPack with each CMake installation) continues to evolve and improve fairly rapidly from release to release.
+Kitware (the company that owns the primary CMake repository and provides the official releases of CMake) puts out three major releases of CMake each year (or about every four months).
```
```suggestion
Kitware (the organization responsible for CMake development) makes three major releases each year (or about every four months).
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+
+The CMake meta-build tool and ecosystem (which includes the tools CTest and CPack with each CMake installation) continues to evolve and improve fairly rapidly from release to release.
+Kitware (the company that owns the primary CMake repository and provides the official releases of CMake) puts out three major releases of CMake each year (or about every four months).
+The best practices for using CMake evolve as new versions of CMake are released.
```
```suggestion
Best practices for using CMake evolve as new versions are released.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+The CMake meta-build tool and ecosystem (which includes the tools CTest and CPack with each CMake installation) continues to evolve and improve fairly rapidly from release to release.
+Kitware (the company that owns the primary CMake repository and provides the official releases of CMake) puts out three major releases of CMake each year (or about every four months).
+The best practices for using CMake evolve as new versions of CMake are released.
+(For example, at the time of this writing, CMake support for a feature called Generator Expressions is still rapidly evolving and can have a dramatic impact on the best way to write CMakeList.txt files.)
```
```suggestion
For example, at the time of this writing, CMake support for [Generator Expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#manual:cmake-generator-expressions(7)) is rapidly evolving and can have a dramatic impact on the best way to write `CMakeList.txt` files.)
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
Not sure I follow. Can't someone just be sure to to use a search that restricts results to nothing less than 6 months or a year old?

bartlettroscoe replied:
I would not say that.  Some CMake problems like how to handle semi-colons ';' chars in the test property ENVIRONMENT are the same now as the were 10+ years ago (and Kitware still does not have a solution to that problem or even document how to successfully handle semi-colons ';' on these properties, even through the workarounds are the same has they have been for many years).  However, advice on how to best generate files in many cases (i.e. using `file(GENERATE ...)` instead of `configure_file(...)`) changed dramatically a few years ago.  And as more commands support generator expressions, the suggested way to solve certain problems will be impacted.

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
What is `this page` referring to here? Can you provide link. Is it the stuff above? Oh, wait, I think you mean this CC article.

```suggestion
The bssw.io editorial team has created this article informed by 13+ years of [advanced CMake usage](https://tribits.org/doc/TribitsDevelopersGuide.html) and hundreds of hours of searching for and reading CMake documentation, especially in the years 2021-2022.
```

bartlettroscoe replied:
> What is `this page` referring to here?

I mean, "the article you are reading".

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
The book contains excellent tutorial material for CMake and yet goes into deep concepts in the CMake language where needed (such as the gotchas and workarounds for dealing with lists and function/macro arguments with data containing quotes `"`, spaces ` ` and semi-colons `;` a frequently misunderstood aspect of CMake).
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
Also, it covers older CMake commands and idioms to avoid and explains why.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
This book addresses the "why" and big picture questions about CMake in a way that no other material has matched.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
It also provides really critical low-level details and tips that are absent in even the [official CMake reference documentation](https://cmake.org/cmake/help/latest/index.html).
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
The author is a significant contributor to CMake and collaborates closely with Kitware staff.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
Not sure this and following sentences are needed. Makes for wordiness but not much in the way of truly needed information.

bartlettroscoe replied:
Fair enough.  The CC article already says that a new edition is put out after each CMake release.  The process by which new editions of the book are created in relation to CMake releases is not really something our readers need to know.  (The reason I mentioned this was to highlight how embedded Craig Scott is with CMake development and releases.)

bartlettroscoe replied:
Addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
Here you refer to the CMake *language* and first line refer to CMake *meta-build system*. Might wanna stick with just one of these. I think *language is fine.

bartlettroscoe replied:
Technically speaking, there is more to CMake than just the CMake language.  There are the executables themselves and their command-line arguments.  There are input files and input file formats that feed into CMake (e.g. JSON formats for certain input files like for presets and test resources).    There are standard CMake modules and their documentation.  And there are many other topics that all fall under "CMake" that are not the CMake language itself.

markcmiller86 replied:
Right but was there a specific reason you used `CMake language` here but used `CMake meta-build` somewhere else (and I think I recall running across `CMake ecosystem` as well)?

Whatever you want to call it (a language, a meta-build system, an ecosystem) is fine. But, I think it should be consistent. And, for what its worth, I think its fine to probably just call it solely by name, `CMake` as in...

- Reference material for CMake including...
- CMake is a large and powerful set of tools...
- ...is still the best advice for the current state of CMake

That avoids entirely any potential cognitive dissidence for what any reader might call it.

bartlettroscoe replied:
I think I can see your point.  However, these adjectives are meant to convey extra meaning.  Here are the definitions:

* Meta-build system: A tool that takes a high-level platform-independent description of a set of software to build and install and then generates the concrete build system (e.g. using Make or Ninja) to actually build the software for a specific system or machine.
* Ecosystem: A collection of related but (in some way) independent software packages or entities.  (In this case CMake, CTest, CPack, CDash and perhaps some related efforts make up the "CMake ecosystem".)
* Language: A set of commands and constructs to execute operations and logic in a computer program.

Are these not well-known definitions?

bartlettroscoe replied:
> Meta-build system

That is actually not consistent with the definition of "meta-build" tool given at:

* https://en.wikipedia.org/wiki/List_of_build_automation_software

But it is consistent with the definition of "meta-build system" which includes CMake at:

* https://sinban04.github.io/build-system/
* http://www.kumarakshay.me/cmake/
* https://alexreinking.com/blog/cmake-is-a-build-system.html

So I think we should remove the term "meta-build system" because there does not seem to be agreement on the definition of a "meta-build system".

bartlettroscoe replied:
Hopefully addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/5fe02d6d33b92c7d64e2a1a9b95d1a18c8f40feb

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
As a developer contributing to a project using CMake (i.e. editing the project's `CMakeLists.txt` files), when consulting the official reference documentation, it is advisable to view documentation for only the minimum version of CMake supported by the project (i.e. for the version `X.Y` listed in the project's top-level `CMakeLists.txt` file calling `cmake_minimum_required(VERSION X.Y)`).
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
For example, if minimum version is 3.17, then one would select that version in the drop-down box (taking you to https://cmake.org/cmake/help/v3.17/).  This helps to avoid accidentally using CMake features available only in newer versions.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```suggestion
Note that the "guides" provided in the online CMake Reference Documentation are also precisely duplicated in the online "Mastering CMake" book.
```

markcmiller86 commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
I would remove this sentence and the next two sentences.

I might add a note that when search on-line sources for CMake documentation, its a good idea to date-restrict the results to material less than one year old.

bartlettroscoe replied:
As I mentioned [above](https://github.com/betterscientificsoftware/bssw.io/pull/1276#discussion_r831594517), you don't want to restrict your web searches to just recent hits but you also want to be careful to make sure this is the best advice given the current state of CMake.

The problem is that the internet contains a lot of material and advice on how to use CMake that is badly out of date given the current state of modern CMake.  But there are also long-standing problems with CMake with workarounds suggested by people 10+ years ago that are still valid today.  So relying on web searches is a double-edge sword.

How can I state this in a compact way?

bartlettroscoe replied:
I shortened this some in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

markcmiller86 replied:
> So relying on web searches is a double-edge sword.

I think I see what you mean. But, I feel like this is true (to one extent or another) about almost *any* HPC software. I don't see what is *unique* about CMake in this regard that requires it being explicitly called out or explicitly alerting readers to the possiblity. Welcome to tech.

bartlettroscoe replied:
That is true.  For example, advice on C++ given 10 years ago related to C++98 may not be the best advance with C++17.  So I guess it goes without saying as languages and tools evolve, what was good advice X years ago is not the best advice today.

So with that I can take out all mentions of web searches.

bartlettroscoe replied:
Hopefully mostly addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/81408104691cb437122181e72f517b02feffe09a

markcmiller86 reviewed (approved):
I think its ok as is but I made a number of suggestions to reduce wordiness in spots. Take 'em or leave 'em

bartlettroscoe commented on CuratedContent/CMakeTutorialAndReferenceResources.md:
```diff
@@ -0,0 +1,87 @@
+# CMake Tutorial and Reference Resources
+
+<!--deck text start-->
+The CMake meta-build system is a large and powerful set of tools to create cross-platform builds from source for C, C++, Fortran and many other compiled programming languages.
```
Okay.  I mention "CMake meta-build system" in the longer introduction paragraph.

bartlettroscoe replied:
Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

bartlettroscoe commented:
@markcmiller86, thanks for your review and all of your great suggestions.  I accepted all of your "suggested changes" and squashed them into the one commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/f4d612c56a244de33624c53d683486e9b67292a9.

And I addressed most of the rest of your suggestions in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3.

There are just a few issues we are not yet on the same page about:

* Advice and caution about web searches (see https://github.com/betterscientificsoftware/bssw.io/pull/1276#discussion_r831565478 and https://github.com/betterscientificsoftware/bssw.io/pull/1276#discussion_r831573772)
* Differentiating between the CMake language itself and all of CMake (see https://github.com/betterscientificsoftware/bssw.io/pull/1276#discussion_r831571638)

If you feel strongly about these, please comment here.  Otherwise, I think I can defend what is currently there (but the text might be improved, especially on the issue of web searches).

I wonder what others (like @bernhold) thinks?  Or this might be getting too much into the weeds to even ask that question.

bartlettroscoe commented:
My last question is:

* Should we keep the "Professional CMake" book cover image? 

It is very large and I am not sure how to make it smaller.  It seems to dominate the article.

For now, I am commenting it out.

bartlettroscoe commented:
@betterscientificsoftware/bssw-maint, I think this is ready for final publication.
